File: python/svix/api/management_authentication.py
# This file is @generated
import typing as t
from dataclasses import dataclass

from .. import models
from ..models import (
    ApiTokenExpireIn,
    ApiTokenIn,
    ApiTokenOut,
    ListResponseApiTokenCensoredOut,
)
from .common import ApiBase, BaseOptions, serialize_params


@dataclass
class ManagementAuthenticationListApiTokensOptions(BaseOptions):
    limit: t.Optional[int] = None
    """Limit the number of returned items"""
    iterator: t.Optional[str] = None
    """The iterator returned from a prior invocation"""
    order: t.Optional[models.Ordering] = None
    """The sorting order of the returned items"""

    def _query_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "limit": self.limit,
                "iterator": self.iterator,
                "order": self.order,
            }
        )


@dataclass
class ManagementAuthenticationCreateApiTokenOptions(BaseOptions):
    idempotency_key: t.Optional[str] = None

    def _header_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "idempotency-key": self.idempotency_key,
            }
        )


@dataclass
class ManagementAuthenticationExpireApiTokenOptions(BaseOptions):
    idempotency_key: t.Optional[str] = None

    def _header_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "idempotency-key": self.idempotency_key,
            }
        )


class ManagementAuthenticationAsync(ApiBase):
    async def list_api_tokens(
        self,
        options: ManagementAuthenticationListApiTokensOptions = ManagementAuthenticationListApiTokensOptions(),
    ) -> ListResponseApiTokenCensoredOut:
        """List all API Tokens."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/management/authentication/api-token",
            path_params={},
            query_params=options._query_params(),
            header_params=options._header_params(),
        )
        return ListResponseApiTokenCensoredOut.model_validate(response.json())

    async def create_api_token(
        self,
        api_token_in: ApiTokenIn,
        options: ManagementAuthenticationCreateApiTokenOptions = ManagementAuthenticationCreateApiTokenOptions(),
    ) -> ApiTokenOut:
        """Create a new API Token."""
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/management/authentication/api-token",
            path_params={},
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=api_token_in.model_dump_json(exclude_unset=True, by_alias=True),
        )
        return ApiTokenOut.model_validate(response.json())

    async def expire_api_token(
        self,
        key_id: str,
        api_token_expire_in: ApiTokenExpireIn,
        options: ManagementAuthenticationExpireApiTokenOptions = ManagementAuthenticationExpireApiTokenOptions(),
    ) -> None:
        """Expire the selected API Token."""
        await self._request_asyncio(
            method="post",
            path="/api/v1/management/authentication/api-token/{key_id}/expire",
            path_params={
                "key_id": key_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=api_token_expire_in.model_dump_json(
                exclude_unset=True, by_alias=True
            ),
        )


class ManagementAuthentication(ApiBase):
    def list_api_tokens(
        self,
        options: ManagementAuthenticationListApiTokensOptions = ManagementAuthenticationListApiTokensOptions(),
    ) -> ListResponseApiTokenCensoredOut:
        """List all API Tokens."""
        response = self._request_sync(
            method="get",
            path="/api/v1/management/authentication/api-token",
            path_params={},
            query_params=options._query_params(),
            header_params=options._header_params(),
        )
        return ListResponseApiTokenCensoredOut.model_validate(response.json())

    def create_api_token(
        self,
        api_token_in: ApiTokenIn,
        options: ManagementAuthenticationCreateApiTokenOptions = ManagementAuthenticationCreateApiTokenOptions(),
    ) -> ApiTokenOut:
        """Create a new API Token."""
        response = self._request_sync(
            method="post",
            path="/api/v1/management/authentication/api-token",
            path_params={},
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=api_token_in.model_dump_json(exclude_unset=True, by_alias=True),
        )
        return ApiTokenOut.model_validate(response.json())

    def expire_api_token(
        self,
        key_id: str,
        api_token_expire_in: ApiTokenExpireIn,
        options: ManagementAuthenticationExpireApiTokenOptions = ManagementAuthenticationExpireApiTokenOptions(),
    ) -> None:
        """Expire the selected API Token."""
        self._request_sync(
            method="post",
            path="/api/v1/management/authentication/api-token/{key_id}/expire",
            path_params={
                "key_id": key_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=api_token_expire_in.model_dump_json(
                exclude_unset=True, by_alias=True
            ),
        )

File: csharp/Svix/ManagementAuthentication.cs
// this file is @generated
#nullable enable
using Microsoft.Extensions.Logging;
using Svix.Models;

namespace Svix
{
    public class ManagementAuthenticationListApiTokensOptions : SvixOptionsBase
    {
        public ulong? Limit { get; set; }
        public string? Iterator { get; set; }
        public Ordering? Order { get; set; }

        public new Dictionary<string, string> QueryParams()
        {
            return SerializeParams(
                new Dictionary<string, object?>
                {
                    { "limit", Limit },
                    { "iterator", Iterator },
                    { "order", Order },
                }
            );
        }
    }

    public class ManagementAuthenticationCreateApiTokenOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class ManagementAuthenticationExpireApiTokenOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class ManagementAuthentication(SvixClient client)
    {
        readonly SvixClient _client = client;

        /// <summary>
        /// List all API Tokens.
        /// </summary>
        public async Task<ListResponseApiTokenCensoredOut> ListApiTokensAsync(
            ManagementAuthenticationListApiTokensOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response =
                    await _client.SvixHttpClient.SendRequestAsync<ListResponseApiTokenCensoredOut>(
                        method: HttpMethod.Get,
                        path: "/api/v1/management/authentication/api-token",
                        queryParams: options?.QueryParams(),
                        headerParams: options?.HeaderParams(),
                        cancellationToken: cancellationToken
                    );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ListApiTokensAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// List all API Tokens.
        /// </summary>
        public ListResponseApiTokenCensoredOut ListApiTokens(
            ManagementAuthenticationListApiTokensOptions? options = null
        )
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ListResponseApiTokenCensoredOut>(
                    method: HttpMethod.Get,
                    path: "/api/v1/management/authentication/api-token",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams()
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ListApiTokens)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new API Token.
        /// </summary>
        public async Task<ApiTokenOut> CreateApiTokenAsync(
            ApiTokenIn apiTokenIn,
            ManagementAuthenticationCreateApiTokenOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            apiTokenIn = apiTokenIn ?? throw new ArgumentNullException(nameof(apiTokenIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<ApiTokenOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/management/authentication/api-token",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: apiTokenIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(CreateApiTokenAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create a new API Token.
        /// </summary>
        public ApiTokenOut CreateApiToken(
            ApiTokenIn apiTokenIn,
            ManagementAuthenticationCreateApiTokenOptions? options = null
        )
        {
            apiTokenIn = apiTokenIn ?? throw new ArgumentNullException(nameof(apiTokenIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ApiTokenOut>(
                    method: HttpMethod.Post,
                    path: "/api/v1/management/authentication/api-token",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: apiTokenIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(CreateApiToken)} failed");

                throw;
            }
        }

        /// <summary>
        /// Expire the selected API Token.
        /// </summary>
        public async Task<bool> ExpireApiTokenAsync(
            string keyId,
            ApiTokenExpireIn apiTokenExpireIn,
            ManagementAuthenticationExpireApiTokenOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            apiTokenExpireIn =
                apiTokenExpireIn ?? throw new ArgumentNullException(nameof(apiTokenExpireIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<bool>(
                    method: HttpMethod.Post,
                    path: "/api/v1/management/authentication/api-token/{key_id}/expire",
                    pathParams: new Dictionary<string, string> { { "key_id", keyId } },
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: apiTokenExpireIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ExpireApiTokenAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Expire the selected API Token.
        /// </summary>
        public bool ExpireApiToken(
            string keyId,
            ApiTokenExpireIn apiTokenExpireIn,
            ManagementAuthenticationExpireApiTokenOptions? options = null
        )
        {
            apiTokenExpireIn =
                apiTokenExpireIn ?? throw new ArgumentNullException(nameof(apiTokenExpireIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<bool>(
                    method: HttpMethod.Post,
                    path: "/api/v1/management/authentication/api-token/{key_id}/expire",
                    pathParams: new Dictionary<string, string> { { "key_id", keyId } },
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: apiTokenExpireIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ExpireApiToken)} failed");

                throw;
            }
        }
    }
}
